lobbyboy/provider.py

```diff
@@ -1,89 +1,148 @@
+import json
+import string
+from abc import ABC, abstractmethod
+from datetime import datetime
 from pathlib import Path
 import logging
 import time
-import threading
+from typing import List, Dict, Callable
 
-from lobbyboy.utils import choose_option
+from paramiko.channel import Channel
+
+from lobbyboy.exceptions import NoAvailableNameException
+from lobbyboy.utils import (
+    confirm_ssh_key_pair,
+    send_to_channel,
+    KeyTypeSupport,
+)
+from lobbyboy.config import LBConfigProvider, LBServerMeta
 
 logger = logging.getLogger(__name__)
+SERVER_FILE = "server.json"
+
 
+class BaseProvider(ABC):
+    def __init__(self, provider_name: str, config: LBConfigProvider, workspace: Path):
+        self.name: str = provider_name
+        self.provider_config: LBConfigProvider = config
+        self.workspace: Path = workspace
 
-class BaseProvider:
-    def __init__(self, provider_name, config, provider_config, data_path):
-        self.config = config
-        self.provider_name = provider_name
-        self.provider_config = provider_config
-        self.data_path = Path(data_path)
+    @staticmethod
+    def time_process_action(c: Channel, act: Callable, max_check: int = 20, interval: int = 3, **action_kws) -> bool:
```

Comment:
can we use the full name (`channel` instead of `c`, and `callback_func` maybe), to consist with other interface functions?

Reply:
yes, of course, it's a little personal hobby, not good😆